Mathlib/Computability/Primrec.lean

```diff
@@ -0,0 +1,1618 @@
+#align primrec.nat_sub Primrec.nat_sub
+
+theorem nat_mul : Primrec₂ ((· * ·) : ℕ → ℕ → ℕ) :=
+  Primrec₂.unpaired'.1 Nat.Primrec.mul
+#align primrec.nat_mul Primrec.nat_mul
+
+theorem cond {c : α → Bool} {f : α → σ} {g : α → σ} (hc : Primrec c) (hf : Primrec f)
+    (hg : Primrec g) : Primrec fun a => bif (c a) then (f a) else (g a) :=
+  (nat_cases (encode_iff.2 hc) hg (hf.comp fst).to₂).of_eq fun a => by cases c a <;> rfl
+#align primrec.cond Primrec.cond
+
+theorem ite {c : α → Prop} [DecidablePred c] {f : α → σ} {g : α → σ} (hc : PrimrecPred c)
+    (hf : Primrec f) (hg : Primrec g) : Primrec fun a => if c a then f a else g a := by
+  simpa using cond hc hf hg
+#align primrec.ite Primrec.ite
+
+theorem nat_le : PrimrecRel ((· ≤ ·) : ℕ → ℕ → Prop) :=
+  (nat_cases nat_sub (const true) (const false).to₂).of_eq fun p => by
+    dsimp [swap]
+    cases' e : p.1 - p.2 with n
+    · simp [tsub_eq_zero_iff_le.1 e]
+    · simp [not_le.2 (Nat.lt_of_sub_eq_succ e)]
+#align primrec.nat_le Primrec.nat_le
+
+theorem nat_min : Primrec₂ (@min ℕ _) :=
+  ite nat_le fst snd
+#align primrec.nat_min Primrec.nat_min
+
+theorem nat_max : Primrec₂ (@max ℕ _) :=
+  ite (nat_le.comp fst snd) snd fst
+#align primrec.nat_max Primrec.nat_max
+
+theorem dom_bool (f : Bool → α) : Primrec f :=
+  (cond .id (const (f true)) (const (f false))).of_eq fun b => by cases b <;> rfl
+#align primrec.dom_bool Primrec.dom_bool
+
+theorem dom_bool₂ (f : Bool → Bool → α) : Primrec₂ f :=
+  (cond fst ((dom_bool (f true)).comp snd) ((dom_bool (f false)).comp snd)).of_eq fun ⟨a, b⟩ => by
+    cases a <;> rfl
+#align primrec.dom_bool₂ Primrec.dom_bool₂
+
+protected theorem not : Primrec not :=
+  dom_bool _
+#align primrec.bnot Primrec.not
+
+protected theorem and : Primrec₂ and :=
+  dom_bool₂ _
+#align primrec.band Primrec.and
+
+protected theorem or : Primrec₂ or :=
+  dom_bool₂ _
+#align primrec.bor Primrec.or
+
+-- porting note: how do we distinguish between propositional and boolean connectives?
+theorem pnot {p : α → Prop} [DecidablePred p] (hp : PrimrecPred p) :
+    PrimrecPred fun a => ¬p a :=
+  (Primrec.not.comp hp).of_eq fun n => by simp
+#align primrec.not Primrec.pnot
+
+/- warning: primrec.and clashes with primrec.band -> Primrec.and
+warning: primrec.and -> Primrec.and is a dubious translation:
+lean 3 declaration is
+  forall {α : Type.{u_1}} [_inst_1 : Primcodable.{u_1} α] {p : α -> Prop} {q : α -> Prop}
+  [_inst_6 : DecidablePred.{succ u_1} α p] [_inst_7 : DecidablePred.{succ u_1} α q],
+  (PrimrecPred.{u_1} α _inst_1 p (fun (a : α) => _inst_6 a)) -> (PrimrecPred.{u_1} α _inst_1 q
+    (fun (a : α) => _inst_7 a)) -> (PrimrecPred.{u_1} α _inst_1 (fun (a : α) => And (p a) (q a))
+    (fun (a : α) => And.decidable (p a) (q a) (_inst_6 a) (_inst_7 a)))
+but is expected to have type
+  Primrec₂.{0, 0, 0} Bool Bool Bool Primcodable.bool Primcodable.bool Primcodable.bool and
+Case conversion may be inaccurate. Consider using '#align primrec.and Primrec.andₓ'. -/
```

Comment:
What's the deal with this warning and the next -- was it taken care of?

Reply:
I'm not sure what the correct name should be. Previously, it was `band` for booleans and `and` for propositions; now, `band` has changed to `and`, but I'm not sure what the Prop version (`and`) should become. Note these are lemmas, not constructors, so they shouldn't be capitalized. I've temporarily used `pand`, `por`, etc. (`p` for "Prop"), but I'll ask on Zulip.

Reply:
Fixed using Eric's suggestion: `pand` moved to namespace `PrimrecPred` (and similarly for others).